ipi_tests/examples/exampletools.py

```diff
@@ -0,0 +1,186 @@
+
+    tree = ET.parse(input_name)
+    root = tree.getroot()
+    clients = list()
+
+    for s, ffsocket in enumerate(root.findall("ffsocket")):
+        name = ffsocket.attrib["name"]
+        ffsocket.attrib["mode"] = driver_info["socket_mode"]
+
+        for element in ffsocket:
+            port = driver_info["port_number"]
+            if element.tag == "port":
+                element.text = str(port)
+            elif element.tag == "address":
+                dd = driver_info["address_name"] + "_" + str(nid) + "_" + str(s)
+                element.text = dd
+                address = dd
+
+        model = driver_info["model"]
+        print("driver:", model)
+        clients.append((model, address, port))
+
+    element = root.find("total_steps")
+    if element is not None:
+        element.text = str(nsteps)
+    else:
+        new_element = ET.SubElement(root, "total_steps")
+        new_element.text = str(nsteps)
+
+    tree.write(open(output_name, "wb"))
+    return clients
+
+
+class Runner_examples(object):
+    def __init__(self, parent, cmd1="i-pi new.xml", cmd2="i-pi-driver"):
+        self.parent = parent
+        self.cmd1 = cmd1
+        self.cmd2 = cmd2
+
+    def run(self, cwd, nid):
+        try:
+            # Create temp file and copy files
+            self.tmp_dir = Path(tempfile.mkdtemp())
+            print("temp folder: {}".format(self.tmp_dir))
+
+            files = os.listdir(self.parent / cwd)
+            for f in files:
+                shutil.copy(self.parent / cwd / f, self.tmp_dir)
+            driver_info = get_driver_info(self.tmp_dir)
+
+            # Modify xml
+            clients = modify_xml(
+                self.tmp_dir / "input.xml", self.tmp_dir / "new.xml", nid, driver_info
+            )
+
+            # Run i-pi
+            ipi = sp.Popen(
+                self.cmd1,
+                cwd=(self.tmp_dir),
+                shell=True,
+                stdout=sp.PIPE,
+                stderr=sp.PIPE,
+            )
+            time.sleep(3)
+
+            # Run drivers
+            driver = list()
+            for client in clients:
+                cmd = self.cmd2 + " -m {} -h {} -u ".format(client[0], client[1])
+                print("cmd:", cmd)
+                driver.append(
+                    sp.Popen(cmd, cwd=(cwd), shell=True, stdout=sp.PIPE, stderr=sp.PIPE)
+                )
+
+            # Check errors
+            ipi_error = ipi.communicate(timeout=45)[1].decode("ascii")
+            if ipi_error != "":
+                print(ipi_error)
+            assert "" == ipi_error
+
+        except sp.TimeoutExpired:
+            raise RuntimeError(
+                "Time is out. Aborted during {} test. \
+              Error {}".format(
+                    str(cwd), ipi.communicate(timeout=2)[0]
+                )
+            )
+
+        except FileNotFoundError:
+            raise ("{}".format(str(cwd)))
+
+        except ValueError:
+            raise ("{}".format(str(cwd)))
```

Comment:
I would like to have a quick Slack chat about this script in the common channel.

Reply:
Sure